sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -0,0 +1,890 @@
+   * responsible for closing the returned batch.
+   *
+   * @param partId the partition id. An exception will be raised up if it is out of range.
+   * @return a Seq of SpillableColumnarBatch if the given "partId" is in range, or null if
+   *         the partition has been released.
+   */
+  def releaseBatchesByPartition(partId: Int): Seq[SpillableColumnarBatch] = {
+    initPartitions()
+    if (0 <= partId && partId < realNumPartitions) {
+      val ret = partsQueue(partId)
+      if (ret != null) {
+        numCurBatches -= ret.length
+      }
+      partsQueue(partId)  = null
+      ret
+    } else {
+      throw new IndexOutOfBoundsException(s"$partId")
+    }
+  }
+
+  override def close(): Unit = {
+    partsQueue.flatten.safeClose()
+    // batches are closed, safe to set all to null
+    (0 until realNumPartitions).foreach(releaseBatchesByPartition)
+  }
+
+  private def initPartitions(): Unit = {
+    if (isNotInited) {
+      partitionBatches()
+      isNotInited = false
+    }
+  }
+
+  private[this] def partitionBatches(): Unit = {
+    while (inputIter.hasNext) {
+      val gpuBatch = inputIter.next()
+      if (gpuBatch.numRows() > 0 && gpuBatch.numCols() > 0) {
+        val types = GpuColumnVector.extractTypes(gpuBatch)
+        // 1) Hash partition on the batch
+        val partedTable = withResource(gpuBatch) { _ =>
+          withResource(GpuColumnVector.from(gpuBatch)) { table =>
+            table.onColumns(keyIndices: _*)
+              .hashPartition(HashType.MURMUR3, realNumPartitions,
+                GpuSubPartitionHashJoin.SUB_PARTITION_HASH_SEED)
+          }
+        }
+        // 2) Split into smaller tables according to partitions
+        val subTables = withResource(partedTable) { _ =>
+          partedTable.getTable.contiguousSplit(partedTable.getPartitions.tail: _*)
+        }
+        // 3) Make each smaller table spillable and cache them in the queue
+        withResource(subTables) { _ =>
+          subTables.zipWithIndex.foreach { case (table, id) =>
+            // skip empty tables
+            if (table.getRowCount > 0) {
+              partsQueue(id) += SpillableColumnarBatch(table, types,
+                SpillPriorities.ACTIVE_ON_DECK_PRIORITY, spillCallback)
+              numCurBatches += 1
+            }
+          }
+        }
+      } else if (gpuBatch.numRows() > 0 && gpuBatch.numCols() == 0) {
+        // Rows only batch
+        // FIXME how to handle this case correctly ?
+        throw new IllegalStateException("Rows-only batch is not allowed")
+      } else {
+        // Skip empty batches
+        gpuBatch.close()
+      }
+    } // end of while
+    numTotalBatches = numCurBatches
+  }
+}
+
+/**
+ * Iterate all the partitions in the input "batchSubPartitioner," and each call to
+ * "next()" will return one or multiple parts as a single "SpillableColumnarBatch",
+ * or None for an empty part, along with its partition id(s).
+ */
+class GpuBatchSubPartitionIterator(
```

Comment:
this iterator feels ripe for unit tests, it would be great to run it through all the scenarios in a scala unit test suite.

Reply:
got it

Reply:
Could we do it as a follow-up?

Reply:
Track it by https://github.com/NVIDIA/spark-rapids/issues/7832

Reply:
Now it is done.